### PR TITLE
update-staging-godeps: only mangle staging repos in staging Godeps.json

### PR DESCRIPTION
**Pull request by sttts**
- this re-adds non-staging k8s.io/* repos to the staging Godeps.jsons
- x-outs instead of removing of staging dependencies in order to get a
  precise trigger for a complete godep restore+save run in the publisher
  bot.

The first breaks k8s.io/kube-aggregator's staging export.

The second potentially leads to inconsistent godeps in our exported staging repos.

### CODE REVIEW

File: hack/update-staging-godeps-dockerized.sh [hunk 1]
```diff
@@ -75,8 +75,21 @@ function updateGodepManifest() {
...
+
+        # x-out staging repo revisions. They will only be known when the publisher bot has created the final export.
+        # We keep the staging dependencies in here though to give the publisher bot a way to detect when the staging
+        # dependencies changed. If they have changed, the bot will run a complete godep restore+save. If they didn't
+        # it will avoid that step, which takes quite some time.
+        jq '.Deps |= map((select(.ImportPath | (startswith("k8s.io/'${repo}'/") or . == "k8s.io/'${repo}'")) | .Rev |= "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx") // .)' > Godeps/Godeps.json.out
+
+      mv Godeps/Godeps.json.out Godeps/Godeps.json
+    done
+    go run ${KUBE_ROOT}/hack/pretty-print-godeps.go < Godeps/Godeps.json > Godeps/Godeps.json.pp
```
Comment (rmmh) [inline hack/update-staging-godeps-dockerized.sh]: Why not `jq . < Godeps/Godeps.json | unexpand --first-only --tabs=2 > Godeps/Godeps.json.pp` ?

Very new JQ supports --tab, but this should be more compatible.

File: hack/update-staging-godeps-dockerized.sh [hunk 2]
```diff
@@ -75,8 +75,21 @@ function updateGodepManifest() {
...
+    # Rewriting Godeps.json to cross-out commits that don't really exist because we haven't pushed the prereqs yet
+    local repo
+    for repo in $(ls -1 ${KUBE_ROOT}/staging/src/k8s.io); do
+      jq '.Deps |= map(.ImportPath |= ltrimstr("k8s.io/kubernetes/staging/src/"))' Godeps/Godeps.json |
+
+        # x-out staging repo revisions. They will only be known when the publisher bot has created the final export.
+        # We keep the staging dependencies in here though to give the publisher bot a way to detect when the staging
+        # dependencies changed. If they have changed, the bot will run a complete godep restore+save. If they didn't
+        # it will avoid that step, which takes quite some time.
+        jq '.Deps |= map((select(.ImportPath | (startswith("k8s.io/'${repo}'/") or . == "k8s.io/'${repo}'")) | .Rev |= "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx") // .)' > Godeps/Godeps.json.out
```
Comment (rmmh) [inline hack/update-staging-godeps-dockerized.sh]: Any particular reason for "x"*40 vs "PENDING" or "MISSING"?

**Comment by rmmh** [comment]
This is a nice cleanup!

File: hack/pretty-print-godeps.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"os"
+)
+
+type Godeps struct {
```
Comment (lavalamp) [inline hack/pretty-print-godeps.go]: Do you actually need this? Isn't it way safer to just unmarshal into map[string]interface?

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json [hunk 4]
```diff
@@ -213,6 +213,10 @@
 		{
 			"ImportPath": "gopkg.in/yaml.v2",
 			"Rev": "53feefa2559fb8dfa8d81baad31be332c97d6c77"
+		},
+		{
+			"ImportPath": "k8s.io/kube-openapi/pkg/common",
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: Didn't see this coming.  @sttts How did apimachinery come to depend on kube-openapi?

File: hack/pretty-print-godeps.go (hunk 3 above)
Comment (sttts) [reply hack/pretty-print-godeps.go]: Tried that. It changes the order. The struct marshaler seems to go field by field.

File: hack/update-staging-godeps-dockerized.sh (hunk 1 above)
Comment (sttts) [reply hack/update-staging-godeps-dockerized.sh]: `--tab` is not in the jq we have. Will try out the unexpand.

File: hack/update-staging-godeps-dockerized.sh (hunk 2 above)
Comment (sttts) [reply hack/update-staging-godeps-dockerized.sh]: No reason other than historic.

File: staging/src/k8s.io/apimachinery/Godeps/Godeps.json (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apimachinery/Godeps/Godeps.json]: One of the imports is for this:

```
func (_ Time) OpenAPIDefinition() openapi.OpenAPIDefinition {
        return openapi.OpenAPIDefinition{
                Schema: spec.Schema{
                        SchemaProps: spec.SchemaProps{
                                Type:   []string{"string"},
                                Format: "date-time",
                        },
                },
        }
}
```

File: hack/pretty-print-godeps.go (hunk 3 above)
Comment (sttts) [reply hack/pretty-print-godeps.go]: with @rmmh's unexpand snipped we can drop this file completely 👍

File: hack/update-staging-godeps-dockerized.sh (hunk 1 above)
Comment (sttts) [reply hack/update-staging-godeps-dockerized.sh]: works perfectly!

**Comment by sttts** [comment]
@rmmh @lavalamp much simpler now, thanks for the feedback! ptal.

**Comment by deads2k** [comment]
lgtm

File: hack/update-staging-godeps-dockerized.sh [hunk 5]
```diff
@@ -75,8 +75,26 @@ function updateGodepManifest() {
...
-    go run "${KUBE_ROOT}/staging/godeps-json-updater.go" --godeps-file="${TMP_GOPATH}/src/k8s.io/${repo}/Godeps/Godeps.json" --override-import-path="k8s.io/${repo}"
+    # Rewriting Godeps.json to cross-out commits that don't really exist because we haven't pushed the prereqs yet
+    local repo
+    for repo in $(ls -1 ${KUBE_ROOT}/staging/src/k8s.io); do
+      # remove staging prefix
+      jq '.Deps |= map(.ImportPath |= ltrimstr("k8s.io/kubernetes/staging/src/"))' Godeps/Godeps.json |
+
+        # x-out staging repo revisions. They will only be known when the publisher bot has created the final export.
+        # We keep the staging dependencies in here though to give the publisher bot a way to detect when the staging
+        # dependencies changed. If they have changed, the bot will run a complete godep restore+save. If they didn't
```
Comment (caesarxuchao) [inline hack/update-staging-godeps-dockerized.sh]: What does "incomplete" mean? Does it mean that the robot will read the commit hash from the local repo in the GOPATH?

**Review by caesarxuchao — COMMENTED** [review]
/lgtm

Thanks!

File: hack/update-staging-godeps-dockerized.sh (hunk 5 above)
Comment (sttts) [reply hack/update-staging-godeps-dockerized.sh]: if you import a package (or some repo) you have not imported before, the Godeps.json before this PR did not change (if the same other package of the same repo was already imported). We only run a complete godep restore+save if the Godeps.json changes (if not, we only replace the staging repo sha1 with a simple sed).

File: hack/update-staging-godeps-dockerized.sh [hunk 6]
```diff
@@ -75,8 +75,26 @@ function updateGodepManifest() {
     rm -rf Godeps # remove the current Godeps.json so we always rebuild it
     GOPATH="${TMP_GOPATH}:${GOPATH}:${GOPATH}/src/k8s.io/kubernetes/staging" godep save ${GODEP_OPTS} ./... 2>&1 | sed 's/^/  /'
 
-    # Rewriting Godeps.json to remove commits that don't really exist because we haven't pushed the prereqs yet
-    go run "${KUBE_ROOT}/staging/godeps-json-updater.go" --godeps-file="${TMP_GOPATH}/src/k8s.io/${repo}/Godeps/Godeps.json" --override-import-path="k8s.io/${repo}"
+    # Rewriting Godeps.json to cross-out commits that don't really exist because we haven't pushed the prereqs yet
+    local repo
+    for repo in $(ls -1 ${KUBE_ROOT}/staging/src/k8s.io); do
+      # remove staging prefix
+      jq '.Deps |= map(.ImportPath |= ltrimstr("k8s.io/kubernetes/staging/src/"))' Godeps/Godeps.json |
```
Comment (smarterclayton) [inline hack/update-staging-godeps-dockerized.sh]: Is this a new dependency?

Comment (sttts) [reply hack/update-staging-godeps-dockerized.sh]: no, it's inside docker and we use jq elsewhere inside the container.

**Review by cblecker — APPROVED** [review]
This looks good to me as well. Better behaved, and 171 less lines of code :)

**Comment by sttts** [comment]
@lavalamp @smarterclayton is this approved?

**Pull request merged**